server/core/src/main/java/org/infinispan/server/core/BackupManager.java

```diff
@@ -0,0 +1,31 @@
+package org.infinispan.server.core;
+
+import java.nio.file.Path;
+import java.util.concurrent.CompletionStage;
+
+import org.infinispan.factories.scopes.Scope;
+import org.infinispan.factories.scopes.Scopes;
+
+/**
+ * Handles all tasks related to the creation/restoration of server backups.
+ *
+ * @author Ryan Emerson
+ * @since 11.0
+ */
+@Scope(Scopes.GLOBAL)
+public interface BackupManager {
+   /**
+    * Create a backup of all containers configured on the server.
+    *
+    * @return a {@link CompletionStage} that on completion returns the {@link Path} to the created backup file.
+    */
+   CompletionStage<Path> create();
+
+   /**
+    * Restore container content from the provided backup bytes.
+    *
+    * @param backup the bytes of the uploaded backup file.
+    * @return a {@link CompletionStage} that completes when all of the entries in the backup have been restored.
+    */
+   CompletionStage<Void> restore(byte[] backup);
```

Comment:
`@param backup the bytes of the uploaded backup file.` <= what is this exactly? is the full data in a single byte{]?

Reply:
This should really be an InputStream or a ReadableByteChannel

Reply:
Sorry I forgot back to get to your original comment Pedro. 

Agreed it shouldn't be byte[], I will update this in the interface. We do have a bigger issue at the moment though, which is AFAIK we can only receive the bytes from a `RestRequest` via `ContentSource#rawContent()` which returns a `byte[]`. As backups will most likely be large, we should support Streams for uploads/downloads with HTTP2, maybe with a fallback of chunked transfer encoding for HTTP/1.1

\cc @gustavonalle

Reply:
Do we really need to send a giant amount of data to the server?  Can't you just accept a ```Path``` to a file somewhere?

If we really need it, it's a great opportunity to expand the REST framework to support a stream of data, using chunked transfer, as it'd be supported both for HTTP/1 and HTTP/2

Reply:
> Do we really need to send a giant amount of data to the server? Can't you just accept a `Path` to a file somewhere?

How do you get the file from the various clients to the server if you don't upload it? The idea is that this will work in Openshift and can be triggered via the Cli and eventually the console.

We could also add the option of the server importing the contents of a local backup file on server startup. In which case the server can pass a `InputStream` of the local file to the `restore` method.

Reply:
> If we really need it, it's a great opportunity to expand the REST framework to support a stream of data, using chunked transfer, as it'd be supported both for HTTP/1 and HTTP/2

I'm not sure if you saw my earlier comment, but my understanding is that chunked transfer was removed in `HTTP/2` https://en.wikipedia.org/wiki/Chunked_transfer_encoding.

Reply:
Our REST server is using HTTP/1 tunnelled in HTTP/2 frames, so it should work.

Reply:
> How do you get the file from the various clients to the server if you don't upload it?

I was just wondering 1 TB backups being sent through HTTP from the client to the server and got a bit worried :)

Reply:
One question: the other way around, i.e., producing the backup, does it first create it on disk and then send it to the client? If so, maybe chunking is not the best but https://en.wikipedia.org/wiki/Byte_serving

Reply:
> does it first create it on disk and then send it to the client

Yep. Currently this is a very simple implementation where it's stored as `staging.zip`, however if we take advantange of byte serving and assign an Id per backup request, we could allow for failed downloads to be resumed. Although we would probably need some form of garbage collection for backups whom's downloads never complete.

Reply:
ok, we can always revisit it later

Reply:
I've created https://issues.redhat.com/browse/ISPN-12145 to track the implementation of chunked transfer encoding.